File: frontend/src/types/receipt.ts
// --- ReceiptItem ---

export interface ReceiptItem {
  id: number;
  name: string | null;
  quantity: number | null;
  price: number | null;
}

export interface ReceiptItemCreate {
  name: string | null;
  quantity: number | null;
  price: number | null;
}

// --- Receipt ---

export interface Receipt {
  id: number;
  store_name: string | null;
  date: string | null;
  total_amount: number | null;
  tax: number | null;
  payment_method: string | null;
  category: string | null;
  image_path: string;
  thumbnail_path: string | null;
  items: ReceiptItem[];
  created_at: string;
  updated_at: string;
}

export interface ReceiptUpdate {
  store_name: string | null;
  date: string | null;
  total_amount: number | null;
  tax: number | null;
  payment_method: string | null;
  category: string | null;
  items: ReceiptItemCreate[];
}

export interface ReceiptListResponse {
  items: Receipt[];
  total: number;
}

// --- Batch Scan ---

export interface BatchScanResultItem {
  filename: string;
  success: boolean;
  receipt: Receipt | null;
  error: string | null;
}

export interface BatchScanResult {
  results: BatchScanResultItem[];
  success_count: number;
  error_count: number;
}

// --- Filter ---

export interface ReceiptFilterParams {
  sort_by?: string;
  sort_order?: "asc" | "desc";
  date_from?: string;
  date_to?: string;
  category?: string;
  amount_min?: number;
  amount_max?: number;
  search?: string;
}

File: backend/app/schemas/receipt.py
from __future__ import annotations

import datetime

from pydantic import BaseModel


# --- ReceiptItem ---

class ReceiptItemBase(BaseModel):
    name: str | None = None
    quantity: float | None = 1
    price: float | None = None


class ReceiptItemCreate(ReceiptItemBase):
    pass


class ReceiptItemResponse(ReceiptItemBase):
    id: int

    model_config = {"from_attributes": True}


# --- Receipt ---

class ReceiptBase(BaseModel):
    store_name: str | None = None
    date: datetime.date | None = None
    total_amount: float | None = None
    tax: float | None = None
    payment_method: str | None = None
    category: str | None = None


class ReceiptUpdate(ReceiptBase):
    items: list[ReceiptItemCreate] = []


class ReceiptResponse(ReceiptBase):
    id: int
    image_path: str
    thumbnail_path: str | None = None
    items: list[ReceiptItemResponse] = []
    created_at: datetime.datetime
    updated_at: datetime.datetime

    model_config = {"from_attributes": True}


class ReceiptListResponse(BaseModel):
    items: list[ReceiptResponse]
    total: int


# --- Vision API ---

class VisionResponse(ReceiptBase):
    items: list[ReceiptItemCreate] = []
